### PR TITLE
[WIP] fit with xshards of pandas df

### PR DESCRIPTION
**Pull request by Le-Zheng**

### CODE REVIEW

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 1]
```diff
@@ -372,6 +372,14 @@ def to_dataset(data, batch_size, batch_per_thread, validation_data,
...
 
     if isinstance(data, SparkXShards):
+        if data._get_class_name() == 'pandas.core.frame.Dataframe':
+            def data_dict_np(df):
+                res = {
+                    "x" : (df[feature_cols[i]].to_numpy() for i in range(len(feature_cols))),
+                    "y" : df[label_cols[0]].to_numpy()
+                }
+                return res
+            data = data.transform_shard(data_dict_np)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:382]: need to handle `validation_data` too

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 2]
```diff
@@ -515,6 +523,14 @@ def fit(self, data,
...
                 "label columns is None; it should not be None in training"
 
+        from zoo.orca.data import SparkXShards
+        if isinstance(data, SparkXShards):
+            if data._get_class_name() == 'pandas.core.frame.Dataframe':
+                assert feature_cols is not None, \
+                    "feature columns is None; it should not be None in training"
+                assert label_cols is not None, \
+                    "label columns is None; it should not be None in training"
+
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:533]: Need to update the python API doc for `fit` above

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 3]
```diff
@@ -523,13 +527,14 @@ def fit(self, data,
...
 
-        from zoo.orca.data import SparkXShards
         if isinstance(data, SparkXShards):
             if data._get_class_name() == 'pandas.core.frame.Dataframe':
                 assert feature_cols is not None, \
                     "feature columns is None; it should not be None in training"
                 assert label_cols is not None, \
                     "label columns is None; it should not be None in training"
+                data = transform_to_shard_dict(data, feature_cols, label_cols)
+                validation_data = transform_to_shard_dict(validation_data, feature_cols, label_cols)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:537]: what if `validation_data` is None? what if `validation_data` is not an XShards of Pandas Dataframe?

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 4]
```diff
@@ -787,6 +812,15 @@ def fit(self, data,
...
                     "either a single tensor or a tuple of tensors"
 
+        if isinstance(data, SparkXShards):
+            if data._get_class_name() == 'pandas.core.frame.Dataframe':
+                assert feature_cols is not None, \
+                    "feature columns is None; it should not be None in training"
+                assert label_cols is not None, \
+                    "label columns is None; it should not be None in training"
+                data = transform_to_shard_dict(data, feature_cols, label_cols)
+                validation_data = transform_to_shard_dict(validation_data, feature_cols, label_cols)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:822]: what if `validation_data` is None? what if `validation_data` is not an XShards of Pandas Dataframe?

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 5]
```diff
@@ -489,13 +490,16 @@ def fit(self, data,
         Train this graph model with train data.
 
         :param data: train data. It can be XShards, Spark DataFrame, tf.data.Dataset.
-        If data is XShards, each partition is a dictionary of  {'x': feature,
-        'y': label}, where feature(label) is a numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be a XShards of Pandas Dataframe or
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:493]: also need to update the python doc for the base `Estimator` class, e.g., https://github.com/intel-analytics/analytics-zoo/blob/e450ba4a5a23f54cbde5e37dcc6d12553addedd6/pyzoo/zoo/orca/learn/tf/estimator.py#L40

File: docs/docs/Orca/orca-tf-estimator.md [hunk 6]
```diff
@@ -97,7 +97,7 @@ fit(data,
 ```
 * `data`: train data. It can be XShards, Spark DataFrame, tf.data.Dataset.
         
-   If `data` is XShards, each element needs to be {'x': a feature numpy array
+   If `data` is XShards, each element can be XShards of Pandas Dataframe or {'x': a feature numpy array
```
Comment (jason-dai) [inline docs/docs/Orca/orca-tf-estimator.md:100]: each element can a Pandas Dataframe or a dictionary ({'x': ..., 'y': ...})

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 7]
```diff
@@ -44,12 +44,15 @@ def fit(self, data, epochs, batch_size=32, feature_cols=None, label_cols=None,
         Train the model with train data.
 
         :param data: train data. It can be XShards, Spark DataFrame, tf.data.Dataset.
-        If data is XShards, each partition is a dictionary of  {'x': feature,
-        'y': label}, where feature(label) is a numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be a XShards of Pandas Dataframe or
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:47]: each element can a Pandas Dataframe or a dictionary of {'x': ..., 'y': ...}

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 8]
```diff
@@ -66,10 +69,11 @@ def predict(self, data, batch_size=4, feature_cols=None, auto_shard_files=False)
         Predict input data
 
         :param data: data to be predicted. It can be XShards, Spark DataFrame.
-        If data is XShards, each partition is a dictionary of  {'x': feature}, where feature is a
-        numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be a XShards of Pandas Dataframe or
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:72]: each partition can a Pandas Dataframe or a dictionary of {'x': ..., 'y': ...}

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 9]
```diff
@@ -88,12 +92,15 @@ def evaluate(self, data, batch_size=32, feature_cols=None, label_cols=None,
         Evaluate model.
 
         :param data: evaluation data. It can be XShards, Spark DataFrame, tf.data.Dataset.
-        If data is XShards, each partition is a dictionary of  {'x': feature, 'y': label}, where
-        feature(label) is a numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be a XShards of Pandas Dataframe or
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:95]: each partition can a Pandas Dataframe or a dictionary of {'x': ..., 'y': ...}

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 10]
```diff
@@ -481,13 +499,16 @@ def fit(self, data,
         Train this graph model with train data.
 
         :param data: train data. It can be XShards, Spark DataFrame, tf.data.Dataset.
-        If data is XShards, each partition is a dictionary of  {'x': feature,
-        'y': label}, where feature(label) is a numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be a XShards of Pandas Dataframe or
+        a dictionary of {'x': feature, 'y': label}, where feature(label) is a numpy array or
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:503]: each partition can a Pandas Dataframe or a dictionary of {'x': ..., 'y': ...}

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 11]
```diff
@@ -572,10 +607,11 @@ def predict(self, data, batch_size=4,
         Predict input data
 
         :param data: data to be predicted. It can be XShards, Spark DataFrame.
-        If data is XShards, each partition is a dictionary of  {'x': feature}, where feature is a
-        numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be XShards of Pandas Dataframe or a dictionary of
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:610]: each partition can a Pandas Dataframe or a dictionary of {'x': ..., 'y': ...}

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 12]
```diff
@@ -625,12 +666,15 @@ def evaluate(self, data, batch_size=32,
         Evaluate model.
 
         :param data: evaluation data. It can be XShards, Spark DataFrame, tf.data.Dataset.
-        If data is XShards, each partition is a dictionary of  {'x': feature, 'y': label}, where
-        feature(label) is a numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be XShards of Pandas Dataframe or
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:669]: each partition can a Pandas Dataframe or a dictionary of {'x': ..., 'y': ...}

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 13]
```diff
@@ -735,13 +787,16 @@ def fit(self, data,
         Train this keras model with train data.
 
         :param data: train data. It can be XShards, Spark DataFrame, tf.data.Dataset.
-        If data is XShards, each partition is a dictionary of  {'x': feature,
-        'y': label}, where feature(label) is a numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be a XShards of Pandas Dataframe  or
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:790]: each partition can a Pandas Dataframe or a dictionary of {'x': ..., 'y': ...}

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 14]
```diff
@@ -817,11 +886,13 @@ def predict(self, data, batch_size=4,
 
         :param data: data to be predicted.
         It can be XShards, Spark DataFrame, or tf.data.Dataset.
-        If data is XShards, each partition is a dictionary of  {'x': feature}, where feature is a
-        numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be XShards of Pandas Dataframe or a dictionary of
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:889]: each partition can a Pandas Dataframe or a dictionary of {'x': ..., 'y': ...}

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 15]
```diff
@@ -865,12 +942,15 @@ def evaluate(self, data, batch_size=32,
         Evaluate model.
 
         :param data: evaluation data. It can be XShards, Spark DataFrame, tf.data.Dataset.
-        If data is XShards, each partition is a dictionary of  {'x': feature, 'y': label}, where
-        feature(label) is a numpy array or a tuple of numpy arrays.
+        If data is XShards, each partition can be XShards of Pandas Dataframe or
+        a dictionary of {'x': feature, 'y': label}, where feature(label) is a numpy array or
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:946]: each partition can a Pandas Dataframe or a dictionary of {'x': ..., 'y': ...}

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 16]
```diff
@@ -515,6 +536,20 @@ def fit(self, data,
             assert label_cols is not None, \
                 "label columns is None; it should not be None in training"
 
+        if isinstance(data, SparkXShards):
```
Comment (yangw1234) [inline pyzoo/zoo/orca/learn/tf/estimator.py:539]: wrap these logics into a utility function in `orca.learn.utils`?

File: pyzoo/zoo/orca/learn/utils.py [hunk 17]
```diff
@@ -228,6 +228,30 @@ def generate_output(feature_lists, label_lists):
...
+    return data
+
+
+def process_xshards_of_pandas_dataframe(data, feature_cols, label_cols=None, validation_data=None,
+                                        mode=None):
+    data = transform_to_shard_dict(data, feature_cols, label_cols)
+    if mode == "fit":
+        if validation_data:
+            assert validation_data._get_class_name() == 'pandas.core.frame.DataFrame',\
+                "train data and validation data should be both Pandas DataFrame"
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/utils.py:248]: train data and validation data should both be XShards of Pandas DataFrame

File: docs/docs/Orca/orca-tf-estimator.md (hunk 6 above)
Comment (jason-dai) [reply docs/docs/Orca/orca-tf-estimator.md:100]: Need to update the above Python doc

**Comment by jason-dai** [comment]
LGTM

**Comment by yangw1234** [comment]
LGTM too. Changes in other estimators will be submitted in another pr?

**Comment by Le-Zheng** [comment]
> LGTM too. Changes in other estimators will be submitted in another pr?

Sure

**Comment by jason-dai** [comment]
> > LGTM too. Changes in other estimators will be submitted in another pr?
> 
> Sure

Yes, let's do that in separate PRs.

**Comment by Le-Zheng** [comment]
Jenkins test http://10.239.47.210:18888/job/ZOO-PR-Validation/5167/
